tpm2/src/util/kmip_util.c

```diff
@@ -0,0 +1,584 @@
+  KeyBlock *key_block = symmetric_key->key_block;
+  KeyValue *key_value = key_block->key_value;
+  ByteString *key_material = key_value->key_material;
+
+  // Set up the official ID and key buffers and clean up.
+  *id = calloc(payload->unique_identifier->size, sizeof(unsigned char));
+  if (*id == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the ID buffer.");
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+  *id_len = payload->unique_identifier->size;
+  memcpy(*id, payload->unique_identifier->value, *id_len);
+
+  *key = calloc(key_material->size, sizeof(unsigned char));
+  if (*key == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the key buffer.");
+    kmyth_clear_and_free(*id, *id_len);
+    *id = NULL;
+    *id_len = 0;
+    kmip_free_response_message(ctx, &message);
+    kmip_set_buffer(ctx, NULL, 0);
+    return 1;
+  }
+  *key_len = key_material->size;
+  memcpy(*key, key_material->value, *key_len);
+
+  kmip_free_response_message(ctx, &message);
+  kmip_set_buffer(ctx, NULL, 0);
+
+  return 0;
+}
+
+//
+// retrieve_key_with_session_key()
+//
+int retrieve_key_with_session_key(int socket_fd,
+                                  unsigned char *session_key,
+                                  size_t session_key_len, unsigned char *key_id,
+                                  size_t key_id_len, unsigned char **key,
+                                  size_t *key_len)
+{
+  KMIP kmip_context = { 0 };
+  kmip_init(&kmip_context, NULL, 0, KMIP_2_0);
+
+  unsigned char *key_request = NULL;
+  size_t key_request_len = 0;
+
+  int result = build_kmip_get_request(&kmip_context,
+                                      key_id, key_id_len,
+                                      &key_request, &key_request_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to build the KMIP Get request.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  unsigned char *encrypted_request = NULL;
+  size_t encrypted_request_len = 0;
+
+  result = aes_gcm_encrypt(session_key, session_key_len,
+                           key_request, key_request_len,
+                           &encrypted_request, &encrypted_request_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to encrypt the KMIP key request.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Sending request for a key with ID: %.*s", key_id_len,
+            key_id);
+  ssize_t write_result =
+    write(socket_fd, encrypted_request, encrypted_request_len);
+
+  if (write_result != encrypted_request_len)
+  {
+    kmyth_log(LOG_ERR, "Failed to fully send the key request.");
+    kmyth_log(LOG_ERR, "Expected to write %zd bytes, only wrote %zd bytes.",
+              encrypted_request_len, write_result);
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  // Read response from B; decrypt with S
+  unsigned char *encrypted_response = calloc(8192, sizeof(unsigned char));
```

Comment:
We should check that the calloc() return value is not NULL.

Reply:
Done